README.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Redis components for Pip.Services in Node.js / ES2017

This module is a part of the [Pip.Services](http://pipservices.org) polyglot microservices toolkit.

The module for working with a Redis database of type key-value contains components for working with cache and locks: RedisCache and RedisLock.

The module contains the following packages:

- **Build** - factory default
- **Cache** - Redis Cache Components
- **Lock** - components of working with locks in Redis

<a name="links"></a> Quick links:

* [Configuration](https://www.pipservices.org/recipies/configuration)
* [API Reference](https://pip-services3-nodex.github.io/pip-services3-redis-nodex/globals.html)
* [Change Log](CHANGELOG.md)
* [Get Help](https://www.pipservices.org/community/help)
* [Contribute](https://www.pipservices.org/community/contribute)

## Use

Install the NPM package as
```bash
npm install pip-services-redis-nodex --save
```

## Develop

For development you shall install the following prerequisites:
* Node.js 8+
* Visual Studio Code or another IDE of your choice
* Docker
* Typescript

Install dependencies:
```bash
npm install
```

Compile the code:
```bash
tsc
```

Run automated tests:
```bash
npm test
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized build and test as:
```bash
./build.ps1
./test.ps1
./clear.ps1
```

## Contacts

The library is created and maintained by **Sergey Seroukhov**.


CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Redis components for Pip.Services in Node.js / ES2017 Changelog

## <a name="1.0.1"></a> 1.0.1 (2021-10-28) 
### Bug fixing
* Fix working with objects, added tests

## <a name="1.0.0"></a> 1.0.0 (2021-05-28) 

### Features
* Added RedisCache
* Added RedisLock



test/lock/RedisLock.test.ts
import { ConfigParams } from 'pip-services3-commons-nodex';

import { RedisLock } from '../../src/lock/RedisLock';
import { LockFixture } from '../fixtures/LockFixture';

suite('RedisLock', ()=> {
    let _lock: RedisLock;
    let _fixture: LockFixture;

    setup(async () => {
        let host = process.env['REDIS_SERVICE_HOST'] || 'localhost';
        let port = process.env['REDIS_SERVICE_PORT'] || 6379;

        _lock = new RedisLock();

        let config = ConfigParams.fromTuples(
            'connection.host', host,
            'connection.port', port
        );
        _lock.configure(config);

        _fixture = new LockFixture(_lock);

        await _lock.open(null);
    });

    teardown(async () => {
        await _lock.close(null);
    });

    test('Try Acquire Lock', async () => {
        await _fixture.testTryAcquireLock();
    });

    test('Acquire Lock', async () => {
        await _fixture.testAcquireLock();
    });

    test('Release Lock', async () => {
        await _fixture.testReleaseLock();
    });

});


test/cache/RedisCache.test.ts
let process = require('process');

import { ConfigParams } from 'pip-services3-commons-nodex';

import { RedisCache } from '../../src/cache/RedisCache';
import { CacheFixture } from '../fixtures/CacheFixture';

suite('RedisCache', ()=> {
    let _cache: RedisCache;
    let _fixture: CacheFixture;

    setup(async () => {
        let host = process.env['REDIS_SERVICE_HOST'] || 'localhost';
        let port = process.env['REDIS_SERVICE_PORT'] || 6379;

        _cache = new RedisCache();

        let config = ConfigParams.fromTuples(
            'connection.host', host,
            'connection.port', port
        );
        _cache.configure(config);

        _fixture = new CacheFixture(_cache);

        await _cache.open(null);
    });

    teardown(async () => {
        await _cache.close(null);
    });

    test('Store and Retrieve', async () => {
        await _fixture.testStoreAndRetrieve();
    });    

    test('Retrieve Expired', async () => {
        await _fixture.testRetrieveExpired();
    });    

    test('Remove', async () => {
        await _fixture.testRemove();
    });    
    
});


test/fixtures/CacheFixture.ts
const assert = require('chai').assert;

import { ICache } from 'pip-services3-components-nodex';

let KEY1: string = "key1";
let KEY2: string = "key2";
let KEY3: string = "key3";
let KEY4: string = "key4";
let KEY5: string = "key5";
let KEY6: string = "key6";

let VALUE1: string = "value1";
let VALUE2 = { val: "value2" };
let VALUE3 = new Date();
let VALUE4 = [1, 2, 3, 4];
let VALUE5 = 12345;
let VALUE6 = null;

export class CacheFixture {
    private _cache: ICache = null;

    public constructor(cache: ICache) {
        this._cache = cache;
    }

    public async testStoreAndRetrieve(): Promise<void> {
        await this._cache.store(null, KEY1, VALUE1, 5000);
        await this._cache.store(null, KEY2, VALUE2, 5000);
        await this._cache.store(null, KEY3, VALUE3, 5000);
        await this._cache.store(null, KEY4, VALUE4, 5000);
        await this._cache.store(null, KEY5, VALUE5, 5000);
        await this._cache.store(null, KEY6, VALUE6, 5000);

        await new Promise(resolve => setTimeout(resolve, 500));

        let val = await this._cache.retrieve(null, KEY1)
        assert.isNotNull(val);
        assert.equal(VALUE1, val);

        val = await this._cache.retrieve(null, KEY2);
        assert.isNotNull(val);
        assert.equal(VALUE2.val, val.val);

        val = await this._cache.retrieve(null, KEY3);
        assert.isNotNull(val);
        assert.equal(VALUE3.toISOString(), val);

        val = await this._cache.retrieve(null, KEY4);
        assert.isNotNull(val);
        assert.lengthOf(val, 4)
        assert.equal(VALUE4[0], val[0]);

        val = await this._cache.retrieve(null, KEY5)
        assert.isNotNull(val);
        assert.equal(VALUE5, val);

        val = await this._cache.retrieve(null, KEY6);
        assert.isNull(val);
    }

    public async testRetrieveExpired(): Promise<void> {
        this._cache.store(null, KEY1, VALUE1, 1000);

        await new Promise(resolve => setTimeout(resolve, 1500));

        let val = await this._cache.retrieve(null, KEY1);
        assert.isNull(val || null);
    }

    public async testRemove(): Promise<void> {
        await this._cache.store(null, KEY1, VALUE1, 1000);

        await this._cache.remove(null, KEY1);

        let val = await this._cache.retrieve(null, KEY1);
        assert.isNull(val || null);
    }

}


src/lock/RedisLock.ts
/** @module lock */
import { ConfigParams } from 'pip-services3-commons-nodex';
import { IConfigurable } from 'pip-services3-commons-nodex';
import { IReferences } from 'pip-services3-commons-nodex';
import { IReferenceable } from 'pip-services3-commons-nodex';
import { IOpenable } from 'pip-services3-commons-nodex';
import { IdGenerator } from 'pip-services3-commons-nodex';
import { InvalidStateException } from 'pip-services3-commons-nodex';
import { ConfigException } from 'pip-services3-commons-nodex';
import { ConnectionParams } from 'pip-services3-components-nodex';
import { ConnectionResolver } from 'pip-services3-components-nodex';
import { CredentialParams } from 'pip-services3-components-nodex';
import { CredentialResolver } from 'pip-services3-components-nodex';
import { Lock } from 'pip-services3-components-nodex';

/**
 * Distributed lock that is implemented based on Redis in-memory database.
 * 
 * ### Configuration parameters ###
 * 
 * - connection(s):           
 *   - discovery_key:         (optional) a key to retrieve the connection from [[https://pip-services3-nodex.github.io/pip-services3-components-nodex/interfaces/connect.idiscovery.html IDiscovery]]
 *   - host:                  host name or IP address
 *   - port:                  port number
 *   - uri:                   resource URI or connection string with all parameters in it
 * - credential(s):
 *   - store_key:             key to retrieve parameters from credential store
 *   - username:              user name (currently is not used)
 *   - password:              user password
 * - options:
 *   - retry_timeout:         timeout in milliseconds to retry lock acquisition. (Default: 100)
 *   - retries:               number of retries (default: 3)
 *  
 * ### References ###
 * 
 * - <code>\*:discovery:\*:\*:1.0</code>        (optional) [[https://pip-services3-nodex.github.io/pip-services3-components-nodex/interfaces/connect.idiscovery.html IDiscovery]] services to resolve connection
 * - <code>\*:credential-store:\*:\*:1.0</code> (optional) Credential stores to resolve credential
 *
 * ### Example ###
 * 
 *     let lock = new RedisRedis();
 *     lock.configure(ConfigParams.fromTuples(
 *       "host", "localhost",
 *       "port", 6379
 *     ));
 * 
 *     lock.open("123");
 * 
 *     await lock.acquire("123", "key1");
 *     try {
 *       // Processing...
 *     } finally {
 *       await lock.releaseLock("123", "key1");
 *     }
 */
export class RedisLock extends Lock implements IConfigurable, IReferenceable, IOpenable {
    private _connectionResolver: ConnectionResolver = new ConnectionResolver();
    private _credentialResolver: CredentialResolver = new CredentialResolver();
    
    private _lock: string = IdGenerator.nextLong();
    private _timeout: number = 30000;
    private _retries: number = 3;

    private _client: any = null;

    /**
     * Configures component by passing configuration parameters.
     * 
     * @param config    configuration parameters to be set.
     */
    public configure(config: ConfigParams): void {
        this._connectionResolver.configure(config);
        this._credentialResolver.configure(config);

        this._timeout = config.getAsIntegerWithDefault('options.timeout', this._timeout);
        this._retries = config.getAsIntegerWithDefault('options.retries', this._retries);
    }

    /**
	 * Sets references to dependent components.
	 * 
	 * @param references 	references to locate the component dependencies. 
     */
    public setReferences(references: IReferences): void {
        this._connectionResolver.setReferences(references);
        this._credentialResolver.setReferences(references);
    }

    /**
	 * Checks if the component is opened.
	 * 
	 * @returns true if the component has been opened and false otherwise.
     */
    public isOpen(): boolean {
        return this._client;
    }

    /**
	 * Opens the component.
	 * 
	 * @param correlationId 	(optional) transaction id to trace execution through call chain.
     */
    public async open(correlationId: string): Promise<void> {
        let connection = await this._connectionResolver.resolve(correlationId);
        if (connection == null) {
            throw new ConfigException(
                correlationId,
                'NO_CONNECTION',
                'Connection is not configured'
            );
        }

        let credential = await this._credentialResolver.lookup(correlationId);

        let options: any = {
            // connect_timeout: this._timeout,
            // max_attempts: this._retries,
            retry_strategy: (options) => { return this.retryStrategy(options); }
        };
                
        if (connection.getUri() != null) {
            options.url = connection.getUri();
        } else {                    
            options.host = connection.getHost() || 'localhost';
            options.port = connection.getPort() || 6379;
        }

        if (credential != null) {
            options.password = credential.getPassword();
        }

        let redis = require('redis');
        this._client = redis.createClient(options);
    }

    /**
	 * Closes component and frees used resources.
	 * 
	 * @param correlationId 	(optional) transaction id to trace execution through call chain.
     */
    public async close(correlationId: string): Promise<void> {
        if (this._client == null) return;

        await new Promise<void>((resolve, reject) => {
            this._client.quit((err) => {
                if (err != null) {
                    reject(err);
                    return;
                }
                resolve();
            });
        });

        this._client = null;    
    }

    private checkOpened(correlationId: string): void {
        if (!this.isOpen()) {
            throw new InvalidStateException(
                correlationId,
                'NOT_OPENED',
                'Connection is not opened'
            );
        }
    }
    
    private retryStrategy(options: any): any {
        if (options.error && options.error.code === 'ECONNREFUSED') {
            // End reconnecting on a specific error and flush all commands with
            // a individual error
            return new Error('The server refused the connection');
        }
        if (options.total_retry_time > this._timeout) {
            // End reconnecting after a specific timeout and flush all commands
            // with a individual error
            return new Error('Retry time exhausted');
        }
        if (options.attempt > this._retries) {
            // End reconnecting with built in error
            return undefined;
        }
        // reconnect after
        return Math.min(options.attempt * 100, 3000);
    }

    /**
     * Makes a single attempt to acquire a lock by its key.
     * It returns immediately a positive or negative result.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a unique lock key to acquire.
     * @param ttl               a lock timeout (time to live) in milliseconds.
     * @returns <code>true</code> if lock was successfully acquired and <code>false</code> otherwise.
     */
    public tryAcquireLock(correlationId: string, key: string, ttl: number): Promise<boolean> {
        this.checkOpened(correlationId);

        return new Promise<boolean>((resolve, reject) => {
            this._client.set(key, this._lock, 'NX', 'PX', ttl, (err, result) => {
                if (err != null) {
                    reject(err);
                    return;
                }
                resolve(result == "OK");
            });    
        });
    }

    /**
     * Releases prevously acquired lock by its key.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a unique lock key to release.
     */
    public releaseLock(correlationId: string, key: string): Promise<void> {
        this.checkOpened(correlationId);

        return new Promise<void>((resolve, reject) => {
            // Start transaction on key
            this._client.watch(key, (err) => {
                if (err != null) {
                    reject(err);
                    return;
                }

                // Read and check if lock is the same
                this._client.get(key, (err, result) => {
                    if (err != null) {
                        reject(err);
                        return;                    
                    }

                    // Remove the lock if it matches
                    if (result == this._lock) {
                        this._client.multi()
                            .del(key)
                            .exec((err) => {
                                if (err != null) reject(err);
                                else resolve();
                            });
                    }
                    // Cancel transaction if it doesn't match
                    else {
                        this._client.unwatch((err) => {
                            if (err != null) reject(err);
                            else resolve();
                        });
                    }
                })
            });
        });
    }    
}

src/cache/RedisCache.ts
/** @module cache */
import { ConfigParams } from 'pip-services3-commons-nodex';
import { IConfigurable } from 'pip-services3-commons-nodex';
import { IReferences } from 'pip-services3-commons-nodex';
import { IReferenceable } from 'pip-services3-commons-nodex';
import { IOpenable } from 'pip-services3-commons-nodex';
import { InvalidStateException } from 'pip-services3-commons-nodex';
import { ConfigException } from 'pip-services3-commons-nodex';
import { ConnectionResolver } from 'pip-services3-components-nodex';
import { CredentialResolver } from 'pip-services3-components-nodex';
import { ICache } from 'pip-services3-components-nodex';

/**
 * Distributed cache that stores values in Redis in-memory database.
 * 
 * ### Configuration parameters ###
 * 
 * - connection(s):           
 *   - discovery_key:         (optional) a key to retrieve the connection from [[https://pip-services3-nodex.github.io/pip-services3-components-nodex/interfaces/connect.idiscovery.html IDiscovery]]
 *   - host:                  host name or IP address
 *   - port:                  port number
 *   - uri:                   resource URI or connection string with all parameters in it
 * - credential(s):
 *   - store_key:             key to retrieve parameters from credential store
 *   - username:              user name (currently is not used)
 *   - password:              user password
 * - options:
 *   - retries:               number of retries (default: 3)
 *   - timeout:               default caching timeout in milliseconds (default: 1 minute)
 *   - max_size:              maximum number of values stored in this cache (default: 1000)        
 *  
 * ### References ###
 * 
 * - <code>\*:discovery:\*:\*:1.0</code>        (optional) [[https://pip-services3-nodex.github.io/pip-services3-components-nodex/interfaces/connect.idiscovery.html IDiscovery]] services to resolve connection
 * - <code>\*:credential-store:\*:\*:1.0</code> (optional) Credential stores to resolve credential
 *
 * ### Example ###
 * 
 *     let cache = new RedisCache();
 *     cache.configure(ConfigParams.fromTuples(
 *       "host", "localhost",
 *       "port", 6379
 *     ));
 * 
 *     await cache.open("123");
 * 
 *     await cache.store("123", "key1", "ABC");
 *     let value = await cache.store("123", "key1"); // Result: "ABC"
 */
export class RedisCache implements ICache, IConfigurable, IReferenceable, IOpenable {
    private _connectionResolver: ConnectionResolver = new ConnectionResolver();
    private _credentialResolver: CredentialResolver = new CredentialResolver();

    private _timeout: number = 30000;
    private _retries: number = 3;

    private _client: any = null;

    /**
     * Creates a new instance of this cache.
     */
    public constructor() { }

    /**
     * Configures component by passing configuration parameters.
     * 
     * @param config    configuration parameters to be set.
     */
    public configure(config: ConfigParams): void {
        this._connectionResolver.configure(config);
        this._credentialResolver.configure(config);

        this._timeout = config.getAsIntegerWithDefault('options.timeout', this._timeout);
        this._retries = config.getAsIntegerWithDefault('options.retries', this._retries);
    }

    /**
     * Sets references to dependent components.
     * 
     * @param references 	references to locate the component dependencies. 
     */
    public setReferences(references: IReferences): void {
        this._connectionResolver.setReferences(references);
        this._credentialResolver.setReferences(references);
    }

    /**
     * Checks if the component is opened.
     * 
     * @returns true if the component has been opened and false otherwise.
     */
    public isOpen(): boolean {
        return this._client;
    }

    /**
     * Opens the component.
     * 
     * @param correlationId 	(optional) transaction id to trace execution through call chain.
     */
    public async open(correlationId: string): Promise<void> {
        let connection = await this._connectionResolver.resolve(correlationId);
        if (connection == null) {
            throw new ConfigException(
                correlationId,
                'NO_CONNECTION',
                'Connection is not configured'
            );
        }

        let credential = await this._credentialResolver.lookup(correlationId);

        let options: any = {
            // connect_timeout: this._timeout,
            // max_attempts: this._retries,
            retry_strategy: (options) => { return this.retryStrategy(options); }
        };

        if (connection.getUri() != null) {
            options.url = connection.getUri();
        } else {
            options.host = connection.getHost() || 'localhost';
            options.port = connection.getPort() || 6379;
        }

        if (credential != null) {
            options.password = credential.getPassword();
        }

        let redis = require('redis');
        this._client = redis.createClient(options);
    }

    /**
     * Closes component and frees used resources.
     * 
     * @param correlationId 	(optional) transaction id to trace execution through call chain.
     */
    public async close(correlationId: string): Promise<void> {
        if (this._client == null) return;

        await new Promise<void>((resolve, reject) => {
            this._client.quit((err) => {
                if (err != null) {
                    reject(err);
                    return;
                }
                resolve();
            });
        });

        this._client = null;
    }

    private checkOpened(correlationId: string): void {
        if (!this.isOpen()) {
            throw new InvalidStateException(
                correlationId,
                'NOT_OPENED',
                'Connection is not opened'
            );
        }
    }

    private retryStrategy(options: any): any {
        if (options.error && options.error.code === 'ECONNREFUSED') {
            // End reconnecting on a specific error and flush all commands with
            // a individual error
            return new Error('The server refused the connection');
        }
        if (options.total_retry_time > this._timeout) {
            // End reconnecting after a specific timeout and flush all commands
            // with a individual error
            return new Error('Retry time exhausted');
        }
        if (options.attempt > this._retries) {
            // End reconnecting with built in error
            return undefined;
        }
        // reconnect after
        return Math.min(options.attempt * 100, 3000);
    }

    /**
     * Retrieves cached value from the cache using its key.
     * If value is missing in the cache or expired it returns null.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a unique value key.
     * @returns a retrieve cached value or <code>null</code> if nothing was found.
     */
    public retrieve(correlationId: string, key: string): Promise<any> {
        this.checkOpened(correlationId);

        return new Promise<any>((resolve, reject) => {
            this._client.get(key, (err, value) => {
                if (err != null) {
                    reject(err);
                    return;
                }
                resolve(value ? JSON.parse(value) : value);
            });
        });
    }

    /**
     * Stores value in the cache with expiration time.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a unique value key.
     * @param value             a value to store.
     * @param timeout           expiration timeout in milliseconds.
     * @returns the stored value.
     */
    public store(correlationId: string, key: string, value: any, timeout: number): Promise<any> {
        this.checkOpened(correlationId);

        return new Promise<any>((resolve, reject) => {
            this._client.set(key, JSON.stringify(value), 'PX', timeout, (err, value) => {
                if (err != null) {
                    reject(err);
                    return;
                }
                resolve(value);
            });
        });
    }

    /**
     * Removes a value from the cache by its key.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a unique value key.
     * @returns the removed value.
     */
    public remove(correlationId: string, key: string): Promise<any> {
        this.checkOpened(correlationId);

        return new Promise<any>((resolve, reject) => {
            this._client.del(key, (err, value) => {
                if (err != null) {
                    reject(err);
                    return;
                }
                resolve(value ? JSON.parse(value) : value);
            });
        });
    }

}